### DISCUSSION TITLE
Is there a way to limit number of rows per user by RLS?

### DISCUSSION THREAD

**Discussion by zccz14** [supabase/supabase#18715] [Questions] [answered]
I created a RLS policy that a user can insert rows with check `user_id = auth.uid()`

So the user can insert any number of rows if the user wants.

But I don't want some user inserting too many rows.

How can I limit a user to insert a certain number of rows?

**Comment by GaryAustin1** [comment, marked as answer]
You would have to add a select in the policy on the table to get the count of rows they have and compare to your limit.

**Reply by zccz14** [reply]
Thanks, [the RLS - Using Joins](https://supabase.com/docs/guides/auth/row-level-security#using-joins) show some cases.

But I am not familiar with policy SQL, any reference for help?

I wrote the policy check expression: (host is a table with columns `id`, `user_id`)

```
(( SELECT count(*) AS count
   FROM host host_1
  WHERE (auth.uid() = host_1.user_id)) < 5)
```

And I got the error message: **infinite recursion detected**

```
{
    "code": "42P17",
    "details": null,
    "hint": null,
    "message": "infinite recursion detected in policy for relation \"host\""
}
```

**Reply by GaryAustin1** [reply]
My guess is you are also making that the select policy.  If you do that it gets your error because you are trying to select the table in the same table you are doing the select query on.  If I'm correct, not sure why you need the select policy to have this limit, just make it something else.
The normal solution for the infinite recursion is to do your extra select in a security definer function so that it bypasses RLS to do the count.

**Reply by zccz14** [reply]
<img width="1137" alt="image" src="https://github.com/supabase/supabase/assets/12707521/57297f83-a5f9-4ec7-b7a8-ced1dba59a49">

you are right.

I want to create a public table `host`, and allow the user to create no more than 5 hosts.

I will create a real service for each host. It has some cost. So I want to limit the number of hosts per user.

**Reply by zccz14** [reply]
In addition, this policy is set for all operation including select, insert, update, and delete.

**Reply by zccz14** [reply]
You mean I should separate the policy by select and others?

**Reply by zccz14** [reply]
Finally, I got the answer: separate the policy. Specify the insert operation by:

<img width="1158" alt="image" src="https://github.com/supabase/supabase/assets/12707521/2978cf83-c947-4c7d-98ab-5dd4caa7703f">
